MA-2247: CourseBlockAPI - return discussion student view data

wajeeha-khalid opened the pull request:
MA-2247: CourseBlockAPI - return student view data for discussion module

wajeeha-khalid commented:
@muzaffaryousaf kindly review

muzaffaryousaf commented on common/lib/xmodule/xmodule/discussion_module.py:
```diff
@@ -123,3 +123,10 @@ def non_editable_metadata_fields(self):
+        """
+        Returns a JSON representation of the student_view of this XModule.
+        """
+        print self.discussion_id
```
Do we need this print statement ?

wajeeha-khalid replied:
Absolutely not; will undo

muzaffaryousaf commented on lms/djangoapps/discussion_api/views.py:
```diff
@@ -158,7 +158,9 @@ class ThreadViewSet(DeveloperErrorViewMixin, ViewSet):
             can be selected.
 
-        The topic_id, text_search, and following parameters are mutually
+        * discussion_id: ID of an inline discussion for any module
```
I think `inline_discussion_id` would make more sense here.

muzaffaryousaf commented on lms/djangoapps/discussion_api/forms.py:
```diff
@@ -37,6 +37,7 @@ class ThreadListGetForm(_PaginationForm):
 
     course_id = CharField()
     topic_id = MultiValueField(required=False)
+    discussion_id = CharField(required=False)
```
`inline_discussion_id`.

muzaffaryousaf commented on lms/djangoapps/discussion_api/tests/test_forms.py:
```diff
@@ -154,6 +155,13 @@ def test_valid_choice_fields(self, field, value):
         self.form_data[field] = value
         self.assert_field_value(field, value)
 
+    def test_discussion_id(self):
```
Doc string missing.

muzaffaryousaf replied:
Can we add new param to `test_mutually_exclusive` test.

muzaffaryousaf commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -274,6 +275,7 @@ def get_thread_list(
         values are "asc" or "desc". The default is "desc".
+    discussion_id: ID of an inline discussion for any module
 
     Note that topic_id_list, text_search, and following are mutually exclusive.
```
We need to specify `inline_discussion_id` here.

muzaffaryousaf commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -274,6 +275,7 @@ def get_thread_list(
         "last_activity_at".
     order_direction: The direction in which to sort the threads by. The only
         values are "asc" or "desc". The default is "desc".
+    discussion_id: ID of an inline discussion for any module
```
Can you add a bit more info. about filtering threads based on this id. The doc string says `Return the list of all discussion threads pertaining to the given course` which is not true now.

wajeeha-khalid replied:
This is true actually because inline discussions are also against course; its only one additional optional filter parameter for this generic endpoint.

muzaffaryousaf commented on lms/djangoapps/discussion_api/tests/test_views.py:
```diff
@@ -519,6 +519,86 @@ def test_order_direction(self, query):
             "sort_order": [query],
         })
 
+    def test_discussion_id(self):
```
Doc string missing.

muzaffaryousaf replied:
I think this test is duplicate of `test_basic` except few values, can we merge this into one ?

muzaffaryousaf commented:
@wajeeha-khalid I'm done with first pass, things are looking good. Let me know when you're done with addressing the feedback.

wajeeha-khalid commented:
jenkins run js

wajeeha-khalid commented:
@muzaffaryousaf @nasthagiri kindly review

wajeeha-khalid commented:
jenkins run js

muzaffaryousaf commented:
@wajeeha-khalid Can you also change PR title and description.

muzaffaryousaf commented:
jenkins run js

symbolist commented:
jenkins run js

nasthagiri commented on common/lib/xmodule/xmodule/discussion_module.py:
```diff
@@ -123,3 +123,9 @@ def non_editable_metadata_fields(self):
+        """
+        Returns a JSON representation of the student_view of this XModule.
+        """
+        return {self.metadata_translations['id']: self.discussion_id}
```
Can you have the dictionary key value be simply "discussion_id"?  The value of this key is a contract between the server and the app and shouldn't be tied to other variables.

nasthagiri commented on lms/djangoapps/course_api/blocks/tests/test_views.py:
```diff
@@ -203,10 +203,14 @@ def test_block_counts_param(self):
+            self.assert_in_iff(
+                'student_view_data',
+                block_data,
+                block_data['type'] == 'video' or block_data['type'] == 'discussion'
```
Assuming we'll add more in the future, can you change this line to:  `block_data['type'] != 'chapter'`

wajeeha-khalid replied:
There are more other blocks than `chapter` where we do not ask for student view data e.g. `videosequence`. In this case `chapter` does not have any student view data even if asked for.
So, `block_data['type'] != 'chapter'` will cause assert failure for all other blocks

nasthagiri replied:
In that case, instead of repeating the list of blocks with `student_view_data` defined, can you create a global/local variable (e.g., `BLOCK_TYPES_WITH_STUDENT_VIEW_DATA`) that lists block types with `student_view_data` (`['video', 'discussion']`)?  And then you can replace this line with `block_data['type'] in BLOCK_TYPES_WITH_STUDENT_VIEW_DATA`.

nasthagiri commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -255,9 +255,11 @@ def get_thread_list(
 ):
     """
-    Return the list of all discussion threads pertaining to the given course
+    Return the list of all discussion threads pertaining to the given course (or if inline_discussion_id is provided
```
The PEP-8 standard for docstring length is 79 characters.  Please wrap the line.

nasthagiri commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -274,6 +276,7 @@ def get_thread_list(
         "last_activity_at".
     order_direction: The direction in which to sort the threads by. The only
         values are "asc" or "desc". The default is "desc".
+    inline_discussion_id: ID of an inline discussion for any module
```
Can we rename this to simply "discussion_id"?  For 2 reasons:
* To use a field name that is consistent with the discussion's `student_view_data` contract.
* In the future, we may have discussions in multiple places, and not necessarily only as inline.

nasthagiri commented:
Please update #2 in this PR's description to say "add a new parameter to the discussion API to query by inline discussion id" rather than "create a new endpoint".

nasthagiri commented on lms/djangoapps/discussion_api/tests/test_views.py:
```diff
@@ -519,6 +528,62 @@ def test_order_direction(self, query):
+            "recursive": ["False"],
+        })
+
+    def test_mutually_exclusive(self):
```
Thanks for adding this test.

nasthagiri commented on lms/djangoapps/discussion_api/views.py:
```diff
@@ -104,7 +104,8 @@ class ThreadViewSet(DeveloperErrorViewMixin, ViewSet):
     **Use Cases**
 
-        Retrieve the list of threads for a course, retrieve thread details,
+        Retrieve the list of threads for a course (or if inline_discussion_id is provided then return
```
No need to add "discussion_id" in the docstring here, since this docstring doesn't specify all the query params.  Rather, the comprehensive list of filter params are documented below.

nasthagiri commented on lms/djangoapps/discussion_api/views.py:
```diff
@@ -158,7 +159,9 @@ class ThreadViewSet(DeveloperErrorViewMixin, ViewSet):
             can be selected.
 
-        The topic_id, text_search, and following parameters are mutually
+        * inline_discussion_id: ID of an inline discussion for any module
```
@catong Any wording recommendation for this API docstring?

catong replied:
@nasthagiri Is this the ID for a content-specific discussion topic that is added via a discussion component? (vs. a course-wide discussion topic)

muzaffaryousaf replied:
@catong Yes, this is content-specific/inline discussion ID not course wide.

catong replied:
Thanks @muzaffaryousaf. @nasthagiri I don't have a lot of context right now but if we are using "inline discussion" throughout, then this string looks fine. My only uncertainty is why we need to say "any module". I would be happy to do a pass through the entire file and take a look at all docstrings. What's the timeline for this PR?

nasthagiri replied:
@catong We aren't bought into the "inline discussion" terminology yet.  We may just use "discussion_id" here to accommodate querying by any discussion identifier.

I don't remember who from the documentation team had reviewed the initial wording in the discussion API last year.  If you'd like to go through the entire file, please do so.  We can make any documentation fixes as part of a separate PR as well, so we don't need to hold up this work (and the subsequent integration testing with the mobile team).

I agree that the "any module" phrase can be updated/removed.

nasthagiri commented:
Do you plan to add an integration-level test that actually creates a discussion module in a course and make sure the API returns its discussion_id?

nasthagiri commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -335,7 +338,10 @@ def get_thread_list(
         query_params["course_id"] = unicode(course.id)
-        query_params["commentable_ids"] = ",".join(topic_id_list) if topic_id_list else None
+        if inline_discussion_id:
+            query_params["commentable_id"] = inline_discussion_id
```
Why reuse the same query param name (commentable_id) for discussion_id?

nasthagiri replied:
Does the comments service already allow filtering by discussion_id?  Or will there be a related PR that updates the comments service?

wajeeha-khalid replied:
Yes, comments service already filters inline discussions and maps content-specific discussion topic id to `commentable_id`.

nasthagiri commented:
@wajeeha-khalid  Completed my pass.  Looks great overall.  I have a few comments, and a question about the comments service.

wajeeha-khalid commented:
jenkins run bokchoy

wajeeha-khalid commented:
@nasthagiri @muzaffaryousaf I have updated it with feedback.

nasthagiri commented on lms/djangoapps/course_api/blocks/tests/test_views.py:
```diff
@@ -30,6 +31,14 @@ def setUpClass(cls):
         cls.course_key = ToyCourseFactory.create().id
         cls.course_usage_key = cls.store.make_course_usage_key(cls.course_key)
+        cls.discussion_id = "discussion_module_id"
+        cls.discussion = ItemFactory.create(
```
Why is it necessary to store `discussion_id` and `discussion` as class variables? You can make `discussion_id` a local variable and remove `cls.discussion`.

nasthagiri commented on lms/djangoapps/course_api/blocks/tests/test_views.py:
```diff
@@ -232,6 +245,18 @@ def test_with_list_field_url(self):
         )
         self.verify_response_with_requested_fields(response)
 
+    def test_discussion_student_view_data(self):
```
This test should be moved to the tests for the discussion module.
The Course Blocks API is generic for all blocks.  We don't want to clutter its tests by verifying `student_view_data` for all block types.

nasthagiri commented on lms/djangoapps/course_api/blocks/tests/test_views.py:
```diff
@@ -82,13 +91,13 @@ def verify_response_block_list(self, response):
         )
 
-    def verify_response_block_dict(self, response):
+    def verify_response_block_dict(self, response, block_usage_key=None):
```
Is the `block_usage_key` optional parameter used?  I don't see a caller passing it in.

nasthagiri commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -274,6 +277,7 @@ def get_thread_list(
         "last_activity_at".
     order_direction: The direction in which to sort the threads by. The only
         values are "asc" or "desc". The default is "desc".
+    discussion_id: Content-specific discussion topic ID
```
Why is this not similar to `topic_id_list` and make it a list of `discussion_id`s?

nasthagiri replied:
By making discussion_id a list, it will allow API clients to get thread information for all (or a bulk of) discussion modules in a course.

wajeeha-khalid replied:
@nimisha In that case, the semantic for 'topic_id_list' is same as 'discussion_id' list. Both 'comments_sevice' and 'DiscussionApi' considers inline discussion ids as 'topic_id/commentable_id'
I would suggest to keep using 'topic_id_list' and remove 'discussion_id' to avoid managing two different fields for same purpose.

nasthagiri replied:
Hmm.. interesting.  I still haven't gotten a strong grasp on the current Discussion terminology.  
The underlying comments service API seems to have a very generic design.  So it may overload multiple concepts into one.  (For example, responses = comments.)  The question is, are "topics" equivalent to "inline discussions" at the outer discussion API layer?

@BenjiLee @aleffert @robrap  What is your take on this?  We want the mobile apps to be able to get all discussion threads filtered by the ID of an inline discussion module.  Can the existing filter, named "topic_id", also apply to inline discussion IDs?  If so, we won't need to change the existing discussion API endpoint.  Otherwise, we'll add a new filter for discussion_ids, but have the underlying implementation equate it to topic_ids.

robrap replied:
@nasthagiri: Are you asking of an "inline discussion" can just be handled via its unique topic, assuming it has a unique topic? Is it possible to get some final help from @jimabramson before he's (your) gone? I will need to be able to answer these questions, but I can't yet with any confidence.

jimabramson replied:
At the data layer (i.e. comments service) each threads points to exactly one topic which is named `commentable_id`.  No distinction is made at that level between an "inline" topic (one that is associated with a specific block inside the course) and a "top-level" topic (a user-defined / user-assigned id and title that are stored as in a dict attribute of the course block itself).

Here's an example of where / how we make the distinction: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/django_comment_client/utils.py#L395

IIRC the distinction is only relevant in the context of access rules (related to cohort config, see that same utils module above for details) or when we are deciding whether or not we should offer a link from this content to the relevant block in the course (i.e., if it's inline, offer a link).

nasthagiri replied:
Thank you @jimabramson and @robrap.  This is very helpful.
The question boiled down to what nomenclature to expose via the discussion API - whether to keep (inline) `discussion_id` as a separate concept from `topic_id`.  Given what you said and the given code, it looks like discussion_ids are _equivalent_ to topic_ids.  The distinction, however, is between a topic being "top-level" or "inline".
Is that correct?

If so, @wajeeha-khalid I agree with you.  We can just use the existing `topic_id_list` as the field for specifying the inline `discussion_id`.  And in so doing, we should change the field name returned by the discussion module's `student_view_data` method to be `topic_id`, so it's clear to clients that they are one and the same.

jimabramson replied:
@nasthagiri to answer your question, i'm pretty sure these are all equivalent: `topic_id` / `discussion_id` / `commentable_id`  (roughly mapping to terminology of discussion API, django comment client, comments service)

naming things...

nasthagiri commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -255,9 +255,12 @@ def get_thread_list(
 ):
     """
     Return the list of all discussion threads pertaining to the given course
+    (or if discussion_id is provided then return threads list against that
```
Can remove this from the docstring here since the comment here is generic.  Not all the possible filters are listed here.  They are listed below.  You can look at `topic_id_list` as an example.

nasthagiri commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -255,9 +255,12 @@ def get_thread_list(
         view=None,
         order_by="last_activity_at",
         order_direction="desc",
+        discussion_id=None,
```
If it doesn't cause a lot of breakage, consider moving the `discussion_id` parameter right below the `topic_id_list` parameter since they are both filters.

wajeeha-khalid commented:
jenkins run lettuce
jenkins run js

wajeeha-khalid commented:
@nasthagiri @muzaffaryousaf I have reverted back the change in DiscussionApi endpoint and added integration test for `student_view_data` for discussion

aleffert commented:
This makes sense to me. I will note that it would be sort of nice to know if the ``topic`` has exactly one child comment when retrieving the block data, since, at least if that is the common case, it might make sense to jump directly to that thread instead of showing a list with one item. But I suppose worst case we can make two API calls in that case.

muzaffaryousaf commented:
jenkins run js

muzaffaryousaf commented:
@nasthagiri Please give it a look, i can address feedback if you have any.

muzaffaryousaf commented:
jenkins run bokchoy

muzaffaryousaf commented:
jenkins run bokchoy

nasthagiri commented on lms/djangoapps/course_api/blocks/tests/test_views.py:
```diff
@@ -203,10 +204,14 @@ def test_block_counts_param(self):
 
     def test_student_view_data_param(self):
-        response = self.verify_response(params={'student_view_data': ['video', 'chapter']})
+        response = self.verify_response(params={'student_view_data': ['video', 'chapter', 'discussion']})
```
Change this to `params={'student_view_data': BLOCK_TYPES_WITH_STUDENT_VIEW_DATA + ['chapter']}`.

nasthagiri commented on lms/djangoapps/discussion_api/tests/test_views.py:
```diff
@@ -235,6 +235,42 @@ def setUp(self):
         self.url = reverse("thread-list")
 
+    def make_expected_thread(self, overrides=None):
+        """ Create a sample expected thread for response
```
Nit: Docstring should be on its own line, as follows:
```
""" 
Create a sample expected thread for response.
""" 
```

nasthagiri commented:
👍  A few minor comments.  Looks good otherwise.

wajeeha-khalid commented:
jenkins run js

muzaffaryousaf commented:
👍 Once the build is green.
